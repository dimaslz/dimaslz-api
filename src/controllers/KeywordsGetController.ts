import { Request, Response } from "express";
import httpStatus from "http-status";

import data from "../data";
import { Controller } from "./Controller";

export default class KeywordsGetController implements Controller {
	async run(req: Request<{ attribute: string }, never, never, never>, res: Response) {
		res.status(httpStatus.OK);

		const jobs = data.jobs;
		const jobsKeywords = jobs.reduce((acc, curr) => [...acc, ...curr.keywords], []);

		const projects = data.projects;
		const projectsKeywords = projects.reduce((acc, curr) => [...acc, ...curr.keywords], []);

		const keywords = [...jobsKeywords, ...projectsKeywords].reduce((acc, curr) => {
			if (!acc.includes(curr)) {
				acc.push(curr);
			}

			return acc;
		}, []);

		return res.json({
			data: keywords,
		});
	}
}
