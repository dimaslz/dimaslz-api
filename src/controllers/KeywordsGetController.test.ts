import httpStatus from "http-status";

import KeywordsGetController from "./KeywordsGetController";

describe("Keywords get controller: GET /keywords", () => {
	it("should return all the project and job keywords by default", async () => {
		const keywordsGetController: KeywordsGetController = new KeywordsGetController();

		const status = jest.fn((v) => v);
		const json = jest.fn((v) => v);
		const result = await keywordsGetController.run.bind(keywordsGetController)(
			{
				params: { attribute: null },
				query: { s: null },
			},
			{
				status,
				json,
			}
		);

		expect(status).toHaveBeenCalledWith(httpStatus.OK);
		expect(result.data.length).toBe(47);

		expect(result).toMatchSnapshot();
	});
});
