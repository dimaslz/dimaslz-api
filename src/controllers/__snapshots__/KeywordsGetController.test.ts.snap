// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keywords get controller: GET /keywords should return all the project and job keywords by default 1`] = `
Object {
  "data": Array [
    "vue2",
    "vue3",
    "vuex",
    "nuxt",
    "jest",
    "testing library",
    "playwright",
    "cypress",
    "SOLID",
    "cloudflare",
    "mentoring",
    "consulting",
    "react",
    "vue",
    "angular",
    "nodejs",
    "docker",
    "infrastructure",
    "side projects",
    "entrepreneurship",
    "socketIO",
    "typescript",
    "vitejs",
    "lidership",
    "nightwatch",
    "testing nodejs",
    "scrum",
    "management",
    "php",
    "jquery",
    "angularjs",
    "npm",
    "ci",
    "javascript",
    "library",
    "components",
    "tailwindcss",
    "preact",
    "tools",
    "api",
    "nextjs",
    "pinia",
    "github actions",
    "developers",
    "svelte",
    "cv",
    "landing",
  ],
}
`;
